Create rule S6772 (#3075)
SonarSource/rspec: rules/S6772/javascript/rule.adoc

== Why is this an issue?

> React JSX differs from the HTML standard in the way it handles newline characters and surrounding whitespace. HTML collapses multiple whitespace characters (including newlines) into a single whitespace, but JSX removes such sequences completely, leaving no space between inline elements separated by the linebreak. This difference in behavior can be confusing and may result in unintended layout, for example, missing whitespace between the link content and the surrounding text.

Reviewer: ```suggestion
React JSX differs from the HTML standard in the way it handles newline characters and surrounding whitespace. HTML collapses multiple whitespace characters (including newlines) into a single whitespace, but JSX removes such sequences completely, leaving no space between inline elements separated by the line break. This difference in behavior can be confusing and may result in unintended layout, for example, missing whitespace between the link content and the surrounding text.
```

To avoid such issues, you should never rely on newline characters in JSX, and explicitly specify whether you want whitespace between inline elements separated by a line break. 

[source,javascript,diff-id=1,diff-type=noncompliant]
----
<div>{/* Noncompliant: ambiguous spacing */}
  Here is some
  <a>space</a>
</div>

<div>{/* Noncompliant: ambiguous spacing */}
  <a>No space</a>
  between these
</div>
----

> To fix the issue either insert an explicit JSX space as a string expression `{' '}`, or insert an empty comments expression `{/*  */}` to indicate that the two parts will be joined together with no space between them.

Reviewer: ```suggestion
To fix the issue, either insert an explicit JSX space as a string expression `{' '}`, or insert an empty comment expression `{/*  */}` to indicate that the two parts will be joined together with no space between them.
```

> [source,text,diff-id=1,diff-type=compliant]

Reviewer: ```suggestion
[source,javascript,diff-id=1,diff-type=compliant]
```
